sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala

```diff
@@ -272,12 +289,36 @@ class SymmetricHashJoinStateManager(
         if (index != numValues - 1) {
           val valuePairAtMaxIndex = keyWithIndexToValue.get(currentKey, numValues - 1)
           if (valuePairAtMaxIndex != null) {
+            // likely case where last element is non-null and we can simply swap with index
             keyWithIndexToValue.put(currentKey, index, valuePairAtMaxIndex.value,
               valuePairAtMaxIndex.matched)
           } else {
-            val projectedKey = getInternalRowOfKeyWithIndex(currentKey)
-            logWarning(s"`keyWithIndexToValue` returns a null value for index ${numValues - 1} " +
-              s"at current key $projectedKey.")
+            // Find the rightmost non null index and swap values with that index,
+            // if index returned is not the same as the passed one
+            val nonNullIndex = getRightMostNonNullIndex(index + 1).getOrElse(index)
+            if (nonNullIndex != index) {
+              val valuePair = keyWithIndexToValue.get(currentKey, nonNullIndex)
+              keyWithIndexToValue.put(currentKey, index, valuePair.value,
+                valuePair.matched)
+            }
+
+            // If nulls were found at the end, get the projected key and log a warning
+            // for the range of null indices.
+            if (nonNullIndex != numValues - 1) {
+              val projectedKey = getInternalRowOfKeyWithIndex(currentKey)
+              logWarning(s"`keyWithIndexToValue` returns a null value for indices " +
+                s"with range from startIndex=${nonNullIndex + 1} " +
+                s"and endIndex=${numValues - 1} " +
+                s"at currentKey=$projectedKey.")
```

Comment:
do we expose user data in the logs here?

Reply:
As discussed, we are only logging the projected join keys along with the range of null value indices. The newly added portion is only the indices. We were logging the projected keys previously too, so no change there:

Here is an example:
```
15:31:47.158 WARN org.apache.spark.sql.execution.streaming.state.SymmetricHashJoinStateManager: `keyWithIndexToValue` returns a null value for indices with range from startIndex=3 and endIndex=4 at currentKey=[false,40,10.0].`

Reply:
but the key itself is effectively data, no?

Reply:
Yes, I believe technically that is true. But without this, we won't have any way of knowing which key(s) we are seeing nulls for. @HeartSaVioR - any thoughts here ? Do you folks think that we should log at DEBUG then or remove altogether ?

Reply:
I generally understand the concern of exposing the data into the log, but we also do it in the other place for tracking down the issue further.

Given this works as a self-fix, it probably depends on the preference/policy about the self-fix (or ignoring the bad one).

Question: Do we want to do it silently without indicating to end users so that they never know the state was broken? Or even with self-fix do we want to alert end users to indicate their state was broken and Spark tries to self-fix? 

If we want to alert, which information end users would want to see? I'm a bit careful that end users may concern if we simply say "there is arbitrary rows being broken and we remove them to correct the state (you don't need to know about the rows we remove)".

Reply:
I unresolved the conversation. Let's try to have a consensus moving forward. If we don't make a consensus sooner, we could leave this as it is, and try to have another discussion thread aiming to address the concern of exposing the data into the log altogether (there is another spot in stream-stream join we log the grouping key.)

cc. @viirya

Reply:
How about we log a hash of the key. Anyway the binary form of the key is unreadable and would need to be programmatically parsed to match the key. We can as well match the key hash.

Reply:
We need to triage "who" we want to guard against seeing this. We will open source the code how to encode the grouping key anyway, so it is pretty trivial for almost everyone to decode the hash if they are basically programmers (and possibly attackers).

Reply:
Yes, the encoding is OSS and even if it was not, you can assume that it is public knowledge. So if we log the binary key you will be able to reconstruct it. I disagree that reconstructing the key from the hash is equivalent - you basically need to know the data in order to match the hash. Any hash should work here.
In general I very strongly suspect that logging data is a security issue.

Reply:
Yeah there are two options, log or not. I don't think obfuscation (which can be decodable by us) would work in anyway.

I generally agree that logging data could be considered as a security issue. If we are enforced to respect the clean and stable security policy then it would have been very clear, but we are yet having such thing on the decision and it is up to the reviewers/committers.

I'm OK with just logging the symptom without context on the data as of now, and even probably need to propose hiding it in other places as well. Probably, this seems to be a strong rationalization of having state data source, at least read functionality; you stop the query if you figure out something is broken, and be able to perform some query against the state to figure out whether state is already broken or not, and if it is broken, which rows are affected.

Reply:
@HeartSaVioR - I think its best to leave this as it is till we have a tool to read/write state as a data source. Once we have that, then we can get rid of these logs. But if you folks feel strongly, then I can just remove the log here.

Reply:
@alex-balikov - Discussed with @HeartSaVioR and we decided to remove the projectedKey for now and retain the warning log. Thx